### ISSUE TITLE
Fib prints wrong output if preceeded by another print

### ISSUE THREAD

**Issue by edwardcwang** [pretis/flexpret#25]
In the following version of fib where something is printed before `fib(16)` is printed, the `fib(16)` print prints `1` instead of the correct value of `987`.

No idea where the bug is.

https://github.com/pretis/flexpret/blob/6bd40ae99589a1000f10258a7cad976a9c79dde5/programs/fib.c#L21

```
int main() {
    // Print an equivalent of "hello world"
    _fp_print(888168);

    const uint32_t x = fib(16);
    _fp_print(x);

    //_fp_print(888168);

    // Terminate the simulation.
    // Put a while loop to make sure no unwanted side effects.
    _fp_finish();
    while(1) {}
    // Not strictly required; just wanted to let the compiler know.
    __builtin_unreachable();
}
```

**Comment by edwardcwang** [comment]
Another example from Shaokai:

```c
#include <stdint.h>
#include <flexpret_io.h>

int main() {
    
    uint32_t x = 1;
    _fp_print(x);
    uint32_t y = 2;
    _fp_print(y); // Prints strange number 0x4E

    // Terminate the simulation.
    // Put a while loop to make sure no unwanted side effects.
    _fp_finish();
    while(1) {}
    // Not strictly required; just wanted to let the compiler know.
    __builtin_unreachable();
}
```

**Comment by lsk567** [comment]
It appears that the issue comes from the `read-under-write` handling in RegisterFile.scala:
https://github.com/pretis/flexpret/blob/0d7c76f3fc6ba8e2f2bda43e9abbe30267d7d8a8/src/main/scala/Core/RegisterFile.scala#L64-L70
Commenting this line out should make the test cases pass.

**Comment by lsk567** [comment]
Also contributing one more test case:
```C
#include <stdint.h>
#include "flexpret_io.h"

#define __print(val) ({ \
    write_csr(CSR_TOHOST, 0xbaaabaaa); \
    write_csr(CSR_TOHOST, val); \
})

uint32_t add(uint32_t a, uint32_t b) {
    return a + b;
}

int main() {
    
    uint32_t x = 1;
    __print(x);

    uint32_t y = 2;
    __print(y);
    
    uint32_t z = add(x, y);
    __print(z);

    // Terminate the simulation.
    // Put a while loop to make sure no unwanted side effects.
    _fp_finish();
    while(1) {}
    // Not strictly required; just wanted to let the compiler know.
    __builtin_unreachable();
}
```
We define `__print` as a macro so that the printing logic doesn't cause an exception, which we noticed can be caused by any function with 2 parameters. In this test case, a `misaligned_load` exception occurs when executing `add a5, a4, a5`.

**Comment by lsk567** [comment]
This bug is addressed with [b6eae0a](https://github.com/pretis/flexpret/commit/b6eae0a3cd764fbb5ef23cc859465a8a7297403a).
